datalad/interface/save.py

```diff
@@ -68,60 +97,82 @@ class Save(Interface):
+        # should be saved
         if auto_add_changes:
             files = [ds.path]
 
-        # treat special case of still untracked subdatasets.
-        # those need to become submodules now, as they are otherwise added
-        # without an entry in .gitmodules, and subsequently break Git's
-        # submodule functionality completely
-        for utf in ds.repo.repo.untracked_files:
-            utf_abspath = opj(ds.path, utf)
-            if not isdir(utf_abspath):
-                # this cannot be a repository
-                continue
-
-            # test whether the potential submodule is scheduled for saving
-            utf_realpath = realpath(utf_abspath)
-            if any([utf_realpath.startswith(_with_sep(realpath(f)))
-                    for f in files]):
-                # matches at least one path -> turn into submodule
-                _install_subds_inplace(
-                    ds=ds,
-                    path=utf_abspath,  # can be ignored, we don't need the return value
-                    relativepath=utf,
-                    name=None)
+        # track whether we modified anything, so it becomes
+        # possible to decide when/what to save further down
+        # and one level up
+        _modified_flag = False
+
+        _modified_flag = untracked_subdatasets_to_submodules(
+            ds, files)
+
+        # now we should have a complete list of submodules to potentially
+        # recurse into
+        if recursive and (recursion_limit is None or recursion_limit):
+            # we only want immediate subdatasets, higher depths will come via
+            # recursion
+            subdss = [Dataset(opj(ds.path, subds_path))
+                      for subds_path in ds.get_subdatasets(
+                          recursive=False)]
+            # skip anything that isn't installed
+            subdss = [d for d in subdss if d.is_installed()]
+            for subds in subdss:
+                subds_modified = Save.__call__(
+                    message=message,
+                    # TODO figure out which files to pass down
+                    # borrow logic from `add`
+                    files=files,
+                    dataset=subds,
+                    auto_add_changes=auto_add_changes,
+                    version_tag=version_tag,
+                    recursive=recursive,
+                    recursion_limit=None if recursion_limit is None else recursion_limit - 1,
+                )
+                if subds_modified:
+                    # stage changes in this submodule
+                    ds.repo.add(relpath(subds.path, ds.path),
+                                git=True)
+                    _modified_flag = True
 
         if files:  # could still be none without auto add changes
             absf = [abspath(f) for f in files]
             # XXX Is there a better way to handle files in mixed repos?
             ds.repo.add(absf)
             ds.repo.add(absf, git=True)
```

Comment:
Re Error: True, fixing this right now while developing tests.

Re always add files: Files are what we are supposed to save, hence they need to be added, no?